src/box/iproto.cc

```diff
@@ -257,15 +255,16 @@ unsigned iproto_readahead = 16320;
 /* The maximal number of iproto messages in fly. */
 static int iproto_msg_max = IPROTO_MSG_MAX_MIN;
 
-const char *
-iproto_bound_address(char *buf)
+char **
+iproto_bound_address(int *size)
 {
-	if (tx_binary.addr_len == 0)
-		return NULL;
-	sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN,
-			  (struct sockaddr *)&tx_binary.addrstorage,
-			  tx_binary.addr_len);
-	return buf;
+	return evio_service_bound_address(&tx_binary, size);
```

Comment:
This is quite a complicated way to return address array you have chosen. Please, drop these mallocs/frees/copying, try this (you will need to add comments and fix potential bugs since I didn't run tests on this):
```Diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index eab6ce825..21869a4f0 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -255,16 +255,19 @@ unsigned iproto_readahead = 16320;
 /* The maximal number of iproto messages in fly. */
 static int iproto_msg_max = IPROTO_MSG_MAX_MIN;
 
-char **
-iproto_bound_address(int *size)
+int
+iproto_addr_count(void)
 {
-	return evio_service_bound_address(&tx_binary, size);
+	return evio_service_count(&tx_binary);
 }
 
-void
-iproto_bound_address_free(char **bound_adress_array)
+const char *
+iproto_addr_str(char *buf, int idx)
 {
-	return evio_service_bound_address_free(bound_adress_array);
+	socklen_t size;
+	const struct sockaddr *addr = evio_service_addr(&tx_binary, idx, &size);
+	sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, addr, size);
+	return buf;
 }
 
 /**
diff --git a/src/box/iproto.h b/src/box/iproto.h
index 99af39882..2afb67495 100644
--- a/src/box/iproto.h
+++ b/src/box/iproto.h
@@ -93,22 +93,11 @@ iproto_thread_stats_get(struct iproto_stats *stats, int thread_id);
 void
 iproto_reset_stat(void);
 
-/**
- * Allocate and return array, which contains representation of the
- * addresses served by iproto. To be shown in box.info. Buffer should
- * be freed by the caller using 'iproto_bound_address_free'. @a size
- * contains count of served addresses. Each address contained in block
- * with SERVICE_NAME_MAXLEN len.
- */
-char **
-iproto_bound_address(int *size);
+int
+iproto_addr_count(void);
 
-/**
- * Free @a bound_address_array allocated by
- * 'iproto_bound_address' function.
- */
-void
-iproto_bound_address_free(char **bound_address_array);
+const char *
+iproto_addr_str(char *buf, int idx);
 
 int
 iproto_rmean_foreach(void *cb, void *cb_ctx);
diff --git a/src/box/lua/info.c b/src/box/lua/info.c
index 17e9603d0..b46e1bebd 100644
--- a/src/box/lua/info.c
+++ b/src/box/lua/info.c
@@ -584,20 +584,21 @@ lbox_info_sql(struct lua_State *L)
 static int
 lbox_info_listen(struct lua_State *L)
 {
-	int size;
-	char **bound_address_array = iproto_bound_address(&size);
-	if (size == 0) {
+	int count = iproto_addr_count();
+	if (count == 0) {
 		lua_pushstring(L, NULL);
-	} else if (size == 1) {
-		lua_pushstring(L, bound_address_array[0]);
-	} else {
-		lua_createtable(L, size, 0);
-		for (int i = 0; i < size; i++) {
-			lua_pushstring(L, bound_address_array[i]);
-			lua_rawseti(L, -2, i + 1);
-		}
+		return 1;
+	}
+	char addrbuf[SERVICE_NAME_MAXLEN];
+	if (count == 1) {
+		lua_pushstring(L, iproto_addr_str(addrbuf, 0));
+		return 1;
+	}
+	lua_createtable(L, count, 0);
+	for (int i = 0; i < count; i++) {
+		lua_pushstring(L, iproto_addr_str(addrbuf, i));
+		lua_rawseti(L, -2, i + 1);
 	}
-	iproto_bound_address_free(bound_address_array);
 	return 1;
 }
 
diff --git a/src/lib/core/evio.c b/src/lib/core/evio.c
index b2d097d93..505a082ae 100644
--- a/src/lib/core/evio.c
+++ b/src/lib/core/evio.c
@@ -63,16 +63,6 @@ evio_service_entry_is_active(struct evio_service_entry *entry)
 	return entry->ev.fd >= 0;
 }
 
-static inline int
-evio_service_entry_bound_address(const struct evio_service_entry *entry,
-				 char *buf)
-{
-	const struct sockaddr *sockaddr = (struct sockaddr *)
-		&entry->addrstorage;
-	return sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, sockaddr,
-				 entry->addr_len);
-}
-
 /** Note: this function does not throw. */
 void
 evio_close(ev_loop *loop, struct ev_io *evio)
@@ -467,30 +457,19 @@ evio_service_reuse_addr(const char **uris, int size)
 	return 0;
 }
 
-char **
-evio_service_bound_address(const struct evio_service *service, int *size)
+int
+evio_service_count(const struct evio_service *service)
 {
-	*size = service->entry_count;
-	if (service->entry_count == 0)
-		return NULL;
-	char **bound_address_array = (char **)
-		xcalloc(service->entry_count,
-			sizeof(char *) + SERVICE_NAME_MAXLEN);
-	char *dstbuf = (char *)bound_address_array +
-		service->entry_count * sizeof(char *);
-	for (int i = 0; i < service->entry_count; i++) {
-		bound_address_array[i] =
-			dstbuf + i * SERVICE_NAME_MAXLEN;
-		evio_service_entry_bound_address(&service->entries[i],
-						 bound_address_array[i]);
-	}
-	return bound_address_array;
+	return service->entry_count;
 }
 
-void
-evio_service_bound_address_free(char **bound_address_array)
+const struct sockaddr *
+evio_service_addr(const struct evio_service *service, int idx, socklen_t *size)
 {
-	free(bound_address_array);
+	assert(idx < service->entry_count);
+	const struct evio_service_entry *e = &service->entries[idx];
+	*size = e->addr_len;
+	return &e->addr;
 }
 
 void
diff --git a/src/lib/core/evio.h b/src/lib/core/evio.h
index 00f612546..6b98664bf 100644
--- a/src/lib/core/evio.h
+++ b/src/lib/core/evio.h
@@ -85,19 +85,11 @@ struct evio_service {
         ev_loop *loop;
 };
 
-/**
- * Return array of strings which represented addresses served by @a service.
- * Save array size in @a size.
- */
-char **
-evio_service_bound_address(const struct evio_service *service, int *size);
+int
+evio_service_count(const struct evio_service *service);
 
-/**
- * Free @a bound_address_array allocated by
- * 'evio_service_bound_address' function.
- */
-void
-evio_service_bound_address_free(char **bound_address_array);
+const struct sockaddr *
+evio_service_addr(const struct evio_service *service, int idx, socklen_t *size);
 
 /** Initialize the service. Don't bind to the port yet. */
 void
```